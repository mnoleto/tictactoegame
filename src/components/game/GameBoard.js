import React, { Component, PropTypes } from 'react';
import GameButton from './GameButton';
import '../../styles/game/GameBoard.scss';

// GameBoard is responsible for render the game board
const GameBoard = ({ board, onClickButton}) => {
  let buttons;
  if(board) {
    buttons = board.map((value, index) => {
      return(
        <GameButton
          key={'button_' + index}
          index={index}
          label={value}
          onClickButton={onClickButton} />
      );
    });
  }

  return (
    <div className="game-board">
      {buttons}
    </div>
  );
}

GameBoard.propTypes = {
  board: PropTypes.array.isRequired,
  onClickButton: PropTypes.func.isRequired
}

export default GameBoard;