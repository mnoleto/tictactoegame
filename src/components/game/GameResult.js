import React, { Component, PropTypes } from 'react';
import '../../styles/game/GameResult.scss';

// GameResult is responsible for render the game result
class GameResult extends Component {
  constructor(props) {
    super(props);
    this.renderMessage = this.renderMessage.bind(this);
  }

  renderMessage() {
    const { players, result } = this.props;
    if(result.status === 'draw') {
      const draw = ['It`s a draw!'];
      return draw.map((value, index) => <span key={'message_' + index}>{value}</span>);
    } else {
      const winnerName = (!result.draw && result.winner === 'X') ? players[0] : players[1];
      const winner = ['The winner is ', <strong>{winnerName}</strong>, '.'];
      return winner.map((value, index) => <span key={'message_' + index}>{value}</span>);
    }
  }

  render() {
    const { onNewGameClick, onNewRoundClick } = this.props;
    
    return (
      <div className="game-result">
        <p>{this.renderMessage()}</p>
        <div className="buttons">
          <button type="button" className="new-round-button" onClick={onNewRoundClick}>New Round</button>
          <button type="button" className="new-game-button" onClick={onNewGameClick}>New Game</button>
        </div>
      </div>
    );
  }
}

GameResult.propTypes = {
  onNewGameClick: PropTypes.func.isRequired,
  onNewRoundClick: PropTypes.func.isRequired,
  players: PropTypes.array.isRequired,
  result: PropTypes.object.isRequired
}

export default GameResult;