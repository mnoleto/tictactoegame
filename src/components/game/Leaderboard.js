import React, { PropTypes } from 'react';
import LeaderboardTr from './LeaderboardTr';
import '../../styles/game/Leaderboard.scss';

/*
 * Stateless component responsible for render the Leaderboard
 */
const Leaderboard = ({ allPlayers }) => {
  let playersItens;
  if(allPlayers && allPlayers.length > 0) {
    // Order players by the number of wins, defeats or draws to show the players with the best scores first
    // 3 points for a win
    // 1 point for a draw
    // -1 point for a defeat
    const orderedPlayers = allPlayers.sort((a, b) => {
      let scoreA = (parseInt(a.wins)*3) + parseInt(a.draws) - parseInt(a.defeats),
        scoreB = (parseInt(b.wins)*3) + parseInt(b.draws) - parseInt(b.defeats);
      return scoreB - scoreA;
    });

    // build an array of table lines
    playersItens = orderedPlayers.map((value, index) => {
      return (
        <LeaderboardTr
          key={'leaderboard_' + index}
          position={index + 1}
          name={value.name}
          wins={value.wins}
          defeats={value.defeats}
          draws={value.draws} />
      );
    });
  }

  return (
    <div className="leaderboard">
      <h4 className="title">Leaderboard</h4>
      <table cellPadding="0" cellSpacing="0">
        <thead>
          <tr className="heading">
            <th colSpan="2" className="name">Name</th>
            <th className="wins">Wins</th>
            <th className="defeats">Defeats</th>
            <th className="draws">Draws</th>
          </tr>
        </thead>
        <tbody>
          {playersItens}
        </tbody>
      </table>
    </div>
  );
};

Leaderboard.propTypes = {
  allPlayers: PropTypes.array
};

export default Leaderboard;