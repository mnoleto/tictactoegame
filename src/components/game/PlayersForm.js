import React, { Component, PropTypes } from 'react';
import PlayerInput from './PlayerInput';
import '../../styles/game/PlayersForm.scss';

// PlayersForm is responsible for render the players form
class PlayersForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      playerX: '',
      playerO: ''
    };

    this.handleInputChange = this.handleInputChange.bind(this);
    this.renderFormFooter = this.renderFormFooter.bind(this);
    this.renderFormHeader = this.renderFormHeader.bind(this);
    this.sendData = this.sendData.bind(this);
  }

  componentWillReceiveProps() {
    if(this.props.players.length > 0 && this.props.players[0] !== this.state.playerX  && this.props.players[1] !== this.state.playerO ) {
      this.setState({
        playerX: this.props.players[0],
        playerO: this.props.players[1]
      });
    }
  }

  handleInputChange(name, value) {
    this.setState({[name]: value});
  }

  sendData() {
    const { onStartGame } = this.props;
    let playerX = this.state.playerX,
      playerO = this.state.playerO;

    onStartGame(playerX, playerO);
  }

  renderFormFooter() {
    const { result, turn } = this.props;
    if(result.status && (result.status && result.status === 'running')) {
      return (
        <div className="turn">It`s <strong>{turn}</strong> turn.</div>
      );
    } else if(!result.status || (result.status && result.status !== 'running')) {
      return (
        <button type="button" className="start-game-button" onClick={this.sendData}>start game</button>
      );
    }
  }

  renderFormHeader() {
    const { result } = this.props;
    if(!result.status || (result.status && result.status !== 'running')) {
      return (
        <legend>Please, inform the players name below:</legend>
      );
    }
    return '';
  }

  render() {
    const { result, turn } = this.props;
    
    return (
      <form className="players-form">
        {this.renderFormHeader()}

        <PlayerInput
          label="X"
          onInputChange={this.handleInputChange}
          player="playerX"
          result={result}
          turn={turn}
          value={this.state.playerX || ''} />

        <PlayerInput
          label="O"
          onInputChange={this.handleInputChange}
          player="playerO"
          result={result}
          turn={turn}
          value={this.state.playerO || ''} />

        {this.renderFormFooter()}
      </form>
    );
  }
}

PlayersForm.propTypes = {
  onStartGame: PropTypes.func.isRequired,
  players: PropTypes.array,
  result: PropTypes.object.isRequired,
  turn: PropTypes.string
};

export default PlayersForm;