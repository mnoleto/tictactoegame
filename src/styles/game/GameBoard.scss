.game-board {
  display: block;
  margin: 0 auto 30px;
  padding: 0 10%;
  width: 100%;
  &:after {
    clear: both;
    content: "";
    display: block;
  }
}

@media (min-width: 768px) {
  .game-board {
    width: 740px;
  }
}

@media (min-width: 992px) {
  .game-board {
    width: 960px;
  }
}

@media (min-width: 1200px) {
  .game-board {
    width: 1170px;
  }
}