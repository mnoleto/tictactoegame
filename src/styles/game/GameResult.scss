@import '../_colors.scss';

.game-result {
  margin-bottom: 50px;
  p {
    font-size: 500%;
    text-align: center;
  }
}

.new-game-button {
  background-color: lighten($blue, 20%);
  border: 0px;
  color: $dark-blue;
  display: block;
  font-size: 100%;
  margin: 15px auto 10px;
  padding: 10px 20px;
}