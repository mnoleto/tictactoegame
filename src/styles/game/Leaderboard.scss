@import '../_colors.scss';

.leaderboard {
  color: $dark-blue;
  display: block;
  margin: 0 auto 30px;
  width: 100%;
  &:after {
    clear: both;
    content: "";
    display: block;
  }
  .title {
    margin-bottom: 5px;
    text-align: center;
  }
  table {
    background: $white;
    margin: 0 auto;
    width: 50%;
  }
  td, th {
    padding: 3px 10px;
    text-align: center;
  }
  tr.heading {
    background-color: lighten($blue, 30%);
  }
  td.name, th.name {
    text-align: left;
  }
  td.position {
    font-weight: 700;
    width: 30px;
  }
}

@media (min-width: 768px) {
  .leaderboard {
    width: 740px;
    table {
      width: 100%;
    }
  }
}

@media (min-width: 992px) {
  .leaderboard {
    width: 960px;
    table {
      width: 70%;
    }
  }
}

@media (min-width: 1200px) {
  .leaderboard {
    width: 1170px;
    table {
      width: 50%;
    }
  }
}