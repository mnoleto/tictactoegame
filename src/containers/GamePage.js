import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PlayersForm from '../components/game/PlayersForm';
import GameBoard from '../components/game/GameBoard';
import GameResult from '../components/game/GameResult';
import Leaderboard from '../components/game/Leaderboard';
import GameCreators from '../actions/game';
import PlayersCreators from '../actions/players';

export class GamePage extends Component {
  constructor(props) {
    super(props);
    
    this.renderContent = this.renderContent.bind(this);
    this.renderLeaderboard = this.renderLeaderboard.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    if(nextProps.game.result && this.props.game.result && nextProps.game.result.status !== this.props.game.result.status) {
      const { fetchPlayers } = this.props.playersActions;
      fetchPlayers();
    }
  }

  componentWillUnmount() {
    const { resetGame } = this.props.gameActions;
    const { fetchPlayers } = this.props.playersActions;
    resetGame();
    fetchPlayers();
  }

  renderContent() {
    const { newGame, newRound, sendPlayersName, registerMove } = this.props.gameActions;
    const { board, players, result, turn } = this.props.game;

    if(result.status === 'finished' || result.status === 'draw') {
      return (
        <GameResult
          onNewGameClick={newGame}
          onNewRoundClick={newRound}
          players={players}
          result={result} />
      );
    } else {
      return (
        <div className="game-running">
          <PlayersForm
            onStartGame={sendPlayersName}
            players={players}
            result={result}
            turn={turn} />
          
          {(result.status === 'running') &&
          <GameBoard
            board={board}
            turn={turn}
            onClickButton={registerMove} />
          }
        </div>
      );
    }
  }

  renderLeaderboard() {
    const { allPlayers } = this.props;

    if(allPlayers && allPlayers.length > 0) {
      return (
        <Leaderboard
          allPlayers={allPlayers} />
      );
    }
  }

  render() {
    return (
      <div className="game-page">
        {this.renderContent()}
        {this.renderLeaderboard()}
      </div>
    );
  }
}

GamePage.propTypes = {
  allPlayers: PropTypes.array,
  game: PropTypes.object,
  gameActions: PropTypes.object,
  playersActions: PropTypes.object
};

function mapStateToProps(state) {
  return {
    allPlayers: state.allPlayers.data,
    game: state.game
  };
}

function mapDispatchToProps(dispatch) {
  return {
    gameActions: bindActionCreators(GameCreators, dispatch), //Turns an object whose values are action creators, into an object with the same keys, but with every action creator wrapped into a dispatch call so they may be invoked directly
    playersActions: bindActionCreators(PlayersCreators, dispatch) 
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(GamePage);
