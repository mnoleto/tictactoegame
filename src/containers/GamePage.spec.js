import React from 'react';
import { mount, shallow } from 'enzyme';
import sinon from 'sinon';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import logger from 'redux-logger';
import createSagaMiddleware from 'redux-saga';

import { GamePage } from './GamePage';
import PlayersForm from '../components/game/PlayersForm';
import GameBoard from '../components/game/GameBoard';
import GameResult from '../components/game/GameResult';
import Leaderboard from '../components/game/Leaderboard';

// create the saga middleware
const sagaMiddleware = createSagaMiddleware();
const middlewares = [
  logger,
  sagaMiddleware,
];

// Use the same middlewares you use with Redux's applyMiddleware
const mockStore = configureMockStore(middlewares);

// Setup the entire state, not just the part Redux passes to the connected component.
const initialState = mockStore({ 
  allPlayers: {},
  game: {},
  routing: {}
});

// GamePage is responsible for render the game grid
describe('<GamePage />', () => {
  let props, mountedGamePage;

  const gamePage = () => {
    if(!mountedGamePage) {
      mountedGamePage = mount(
        <Provider store={initialState}>
          <GamePage {...props} />
        </Provider>
      );
    }
    return mountedGamePage;
  };

  beforeEach(() => {
    props = {
      allPlayers: [],
      game: {
        board: ['E', 'E', 'E', 'E', 'E', 'E', 'E', 'E', 'E', ],
        players: [],
        result: {status: 'running'},
        turn: ''
      },
      gameActions: {
        newGame: jest.fn(),
        newRound: jest.fn(), 
        registerMove: jest.fn(),
        resetGame: jest.fn(),
        sendPlayersName: jest.fn()
      },
      playersActions: {
        fetchPlayers: jest.fn()
      }
    };
    mountedGamePage = undefined;
  });

  describe('Test methods', () => {
    let gamePageRenderer;
    beforeEach(() => {
      props.playersActions.fetchPlayers = sinon.spy();
      props.gameActions.resetGame = sinon.spy();
      gamePageRenderer = shallow(<GamePage {...props} />);
    });

    describe('componentWillReceiveProps', () => {
      it('should fetch players', () => {
        gamePageRenderer.instance().componentWillReceiveProps({
          game: {
            board: ['E', 'E', 'E', 'E', 'E', 'E', 'E', 'E', 'E', ],
            players: [],
            result: {status: 'finished', winner: 'X'},
            turn: ''
          }
        });
        expect(props.playersActions.fetchPlayers.callCount).toEqual(1);
      });
    });

    describe('componentWillUnmount', () => {
      it('should reset the game and fetch players when componentWillUnmount', () => {
        gamePageRenderer.instance().componentWillUnmount();
        expect(props.gameActions.resetGame.callCount).toEqual(1);
        expect(props.playersActions.fetchPlayers.callCount).toEqual(1);
      });
    });
  });

  describe('render', () => {
    it('render the GamePage component', () => {
      const gamePageRenderer = shallow(<GamePage {...props} />);
      expect(gamePageRenderer).toMatchSnapshot();
    });

    it('render a div as root element', () => {
      const gamePageElement = gamePage().find('div');
      const wrappingDiv = gamePageElement.first();
      expect(wrappingDiv.children()).toEqual(gamePage().children());
    });

    it('should wrapper has the class game-page', () => {
      const gamePageElement = gamePage().find('div');
      const wrappingDiv = gamePageElement.first();
      expect(wrappingDiv.hasClass('game-page')).toBeTruthy();
    });

    describe('renderContent()', () => {
      it('should render the players form if status is not equal finished or draw', () => {
        let gamePageRenderer = gamePage().find(GamePage);
        let playersForm = gamePageRenderer.find(PlayersForm);
        expect(playersForm.length).toEqual(1);

        props.game.result = {status: 'waiting'};
        gamePageRenderer = gamePage().find(GamePage);
        playersForm = gamePageRenderer.find(PlayersForm);
        expect(playersForm.length).toEqual(1);
      });

      it('should not render the players form if status is finished or draw', () => {
        props.game.result = {status: 'finished'};
        let gamePageRenderer = gamePage().find(GamePage);
        let playersForm = gamePageRenderer.find(PlayersForm);
        expect(playersForm.length).toEqual(0);

        props.game.result = {status: 'draw'};
        gamePageRenderer = gamePage().find(GamePage);
        playersForm = gamePageRenderer.find(PlayersForm);
        expect(playersForm.length).toEqual(0);
      });

      it('should render the game board if status is equal to running', () => {
        props.game.result = {status: 'running'};
        const gamePageRenderer = gamePage().find(GamePage);
        const gameBoard = gamePageRenderer.find(GameBoard);
        expect(gameBoard.length).toEqual(1);
      });

      it('should not render the game board if status is not running', () => {
        props.game.result = {status: 'finished'};
        let gamePageRenderer = gamePage().find(GamePage);
        let gameBoard = gamePageRenderer.find(GameBoard);
        expect(gameBoard.length).toEqual(0);

        props.game.result = {status: 'draw'};
        gamePageRenderer = gamePage().find(GamePage);
        gameBoard = gamePageRenderer.find(GameBoard);
        expect(gameBoard.length).toEqual(0);

        props.game.result = {status: 'waiting'};
        gamePageRenderer = gamePage().find(GamePage);
        gameBoard = gamePageRenderer.find(GameBoard);
        expect(gameBoard.length).toEqual(0);
      });

      it('should render the game result if status is equal finished or draw', () => {
        props.game.result = {status: 'finished'};
        let gamePageRenderer = gamePage().find(GamePage);
        let gameResult = gamePageRenderer.find(GameResult);
        expect(gameResult.length).toEqual(1);

        props.game.result = {status: 'draw'};
        gamePageRenderer = gamePage().find(GamePage);
        gameResult = gamePageRenderer.find(GameResult);
        expect(gameResult.length).toEqual(1);
      });

      it('should not render the game result if status is not finished or draw', () => {
        props.game.result = {status: 'running'};
        let gamePageRenderer = gamePage().find(GamePage);
        let gameResult = gamePageRenderer.find(GameResult);
        expect(gameResult.length).toEqual(0);

        props.game.result = {status: 'waiting'};
        gamePageRenderer = gamePage().find(GamePage);
        gameResult = gamePageRenderer.find(GameResult);
        expect(gameResult.length).toEqual(0);
      });
    });

    describe('renderLeaderboard()', () => {
      it('should not render the leaderboard if allPlayers length is equal 0', () => {
        const gamePageRenderer = gamePage().find(GamePage);
        const leaderboard = gamePageRenderer.find(Leaderboard);
        expect(leaderboard.length).toEqual(0);
      });

      it('should render the leaderboard if allPlayers length is higher than 0', () => {
        props.allPlayers = [{name: 'Player 1', wins: 0, defeats: 0, draws: 0}];
        const gamePageRenderer = gamePage().find(GamePage);
        const leaderboard = gamePageRenderer.find(Leaderboard);
        expect(leaderboard.length).toEqual(1);
      });
    });
  });
});