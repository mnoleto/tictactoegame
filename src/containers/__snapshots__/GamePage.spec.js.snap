exports[`<GamePage /> render render the GamePage component 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="game-page">
    <div
        className="game-running">
        <PlayersForm
            onStartGame={[Function]}
            players={Array []}
            result={
                Object {
                    "status": "running",
                  }
            }
            turn="" />
        <GameBoard
            board={
                Array [
                    "E",
                    "E",
                    "E",
                    "E",
                    "E",
                    "E",
                    "E",
                    "E",
                    "E",
                  ]
            }
            onClickButton={[Function]}
            turn="" />
    </div>
</div>,
  "nodes": Array [
    <div
      className="game-page">
      <div
            className="game-running">
            <PlayersForm
                  onStartGame={[Function]}
                  players={Array []}
                  result={
                        Object {
                              "status": "running",
                            }
                  }
                  turn="" />
            <GameBoard
                  board={
                        Array [
                              "E",
                              "E",
                              "E",
                              "E",
                              "E",
                              "E",
                              "E",
                              "E",
                              "E",
                            ]
                  }
                  onClickButton={[Function]}
                  turn="" />
      </div>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <GamePage
        allPlayers={Array []}
        game={
                Object {
                        "board": Array [
                          "E",
                          "E",
                          "E",
                          "E",
                          "E",
                          "E",
                          "E",
                          "E",
                          "E",
                        ],
                        "players": Array [],
                        "result": Object {
                          "status": "running",
                        },
                        "turn": "",
                      }
        }
        gameActions={
                Object {
                        "newGame": [Function],
                        "newRound": [Function],
                        "registerMove": [Function],
                        "resetGame": [Function],
                        "sendPlayersName": [Function],
                      }
        }
        playersActions={
                Object {
                        "fetchPlayers": [Function],
                      }
        } />,
      "_debugID": 5,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": GamePage {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "allPlayers": Array [],
          "game": Object {
            "board": Array [
              "E",
              "E",
              "E",
              "E",
              "E",
              "E",
              "E",
              "E",
              "E",
            ],
            "players": Array [],
            "result": Object {
              "status": "running",
            },
            "turn": "",
          },
          "gameActions": Object {
            "newGame": [Function],
            "newRound": [Function],
            "registerMove": [Function],
            "resetGame": [Function],
            "sendPlayersName": [Function],
          },
          "playersActions": Object {
            "fetchPlayers": [Function],
          },
        },
        "refs": Object {},
        "renderContent": [Function],
        "renderLeaderboard": [Function],
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 3,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="game-page">
          <div
                    className="game-running">
                    <PlayersForm
                              onStartGame={[Function]}
                              players={Array []}
                              result={
                                        Object {
                                                  "status": "running",
                                                }
                              }
                              turn="" />
                    <GameBoard
                              board={
                                        Array [
                                                  "E",
                                                  "E",
                                                  "E",
                                                  "E",
                                                  "E",
                                                  "E",
                                                  "E",
                                                  "E",
                                                  "E",
                                                ]
                              }
                              onClickButton={[Function]}
                              turn="" />
          </div>
</div>,
        "_debugID": 6,
        "_renderedOutput": <div
          className="game-page">
          <div
                    className="game-running">
                    <PlayersForm
                              onStartGame={[Function]}
                              players={Array []}
                              result={
                                        Object {
                                                  "status": "running",
                                                }
                              }
                              turn="" />
                    <GameBoard
                              board={
                                        Array [
                                                  "E",
                                                  "E",
                                                  "E",
                                                  "E",
                                                  "E",
                                                  "E",
                                                  "E",
                                                  "E",
                                                  "E",
                                                ]
                              }
                              onClickButton={[Function]}
                              turn="" />
          </div>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <GamePage
    allPlayers={Array []}
    game={
        Object {
            "board": Array [
              "E",
              "E",
              "E",
              "E",
              "E",
              "E",
              "E",
              "E",
              "E",
            ],
            "players": Array [],
            "result": Object {
              "status": "running",
            },
            "turn": "",
          }
    }
    gameActions={
        Object {
            "newGame": [Function],
            "newRound": [Function],
            "registerMove": [Function],
            "resetGame": [Function],
            "sendPlayersName": [Function],
          }
    }
    playersActions={
        Object {
            "fetchPlayers": [Function],
          }
    } />,
}
`;
